Name: Joseph Ridge Ogada
Adm.: 166895
task: Assignment 2

# Importing relevant *packages*

In [1]:
import pandas as pd
import numpy as np 
# import matplotlib.pyplot as plt 
# import seaborn as sns 

ImportError: DLL load failed while importing _cext: The specified module could not be found.

# Business Understanding

- Dataset source:  dataset from the MIMIC-III database
- Contains information on in-hospital mortality from the monitoring of patients in the Intensive Care
Unit (ICU) for 48 hours

# Data Understanding

#### Feature Description

| Feature                         | Description                                                                 |
|-------------------------------------|-----------------------------------------------------------------------------|
| Capillary refill rate               | The time it takes for color to return to an external capillary bed (such as a fingertip) after pressure is applied. It is an indicator of peripheral perfusion. |
| Diastolic Blood Pressure            | The pressure in the arteries when the heart rests between beats. It's the lower of the two blood pressure measurements. |
| Fraction inspired oxygen            | The concentration of oxygen in the air mixture that is inhaled by the patient, often measured in ventilated patients. |
| Glascow coma scale eye opening      | A component of the Glasgow Coma Scale (GCS) that measures a patient's ability to open their eyes in response to stimuli. |
| Glascow coma scale motor response    | A component of the GCS that assesses a patient's motor response to stimuli, indicating brain function. |
| Glascow coma scale total            | The total score of the Glasgow Coma Scale, which assesses the level of consciousness in a person following a traumatic brain injury. |
| Glascow coma scale verbal response   | A component of the GCS that evaluates a patient’s verbal response, indicating their level of consciousness. |
| Glucose                             | The level of sugar (glucose) in the blood, an important indicator of metabolic status. |
| Heart Rate                          | The number of heartbeats per minute, indicating the functioning of the cardiovascular system. |
| Height                              | The measurement of a patient's stature, typically recorded in centimeters or meters. |
| Mean blood pressure                 | The average pressure in a patient's arteries during one cardiac cycle, indicating overall blood pressure. |
| Oxygen saturation                   | The percentage of hemoglobin binding sites in the bloodstream occupied by oxygen, indicating the oxygenation status of the patient. |
| Respiratory rate                    | The number of breaths taken per minute, indicating the patient’s respiratory status. |
| Systolic blood pressure             | The pressure in the arteries when the heart beats. It is the higher of the two blood pressure measurements. |
| Temperature                         | The body temperature of the patient, indicating metabolic and homeostatic status. |
| Weight                              | The body mass of the patient, typically recorded in kilograms. |
| pH                                  | The measure of acidity or alkalinity of the blood, indicating the balance of acids and bases in the body. |
| Patient_id                          | A unique identifier assigned to each patient for tracking and reference purposes. |
| target                              | The outcome or variable of interest (mortality), which the model is aiming to predict. |
 

####  Relevant Functions

In [15]:
# function to get types and shape of data

#  Checking for Outliers
def  checkNullValues(df):
    print("Null Values Present in Data")
    print(df.isna().sum() * 100 / len(df))



####  Read Data: 

In [23]:
df  = pd.read_csv("ihm_48_hours.csv")
df.head()

Capillary refill rate  Diastolic blood pressure  Fraction inspired oxygen  \
0                    NaN                      73.0                       NaN   
1                    NaN                      73.0                       NaN   
2                    NaN                      73.0                       NaN   
3                    NaN                      73.0                       NaN   
4                    NaN                      73.0                       NaN   

  Glascow coma scale eye opening Glascow coma scale motor response  \
0                  Spontaneously                    Obeys Commands   
1                  Spontaneously                    Obeys Commands   
2                  Spontaneously                    Obeys Commands   
3                  Spontaneously                    Obeys Commands   
4                  Spontaneously                    Obeys Commands   

   Glascow coma scale total Glascow coma scale verbal response     Glucose  \
0                       NaN                           Oriented  -11.396037   
1                       NaN                           Oriented  115.000000   
2                       NaN                           Oriented  115.000000   
3                       NaN                           Oriented  115.000000   
4                       NaN                           Oriented  115.000000   

   Heart Rate  Height  Mean blood pressure  Oxygen saturation  \
0  -19.976803     NaN                 76.0          94.000000   
1   96.000000     NaN                 76.0          95.000000   
2   96.000000     NaN                 76.0          -6.497052   
3   96.000000     NaN                 76.0          95.000000   
4   96.000000     NaN                 76.0          95.000000   

   Respiratory rate  Systolic blood pressure  Temperature  Weight  pH  \
0              17.0                    116.0    36.388889    83.5 NaN   
1              18.0                    116.0    36.388889    83.5 NaN   
2              18.0                    116.0    36.388889    83.5 NaN   
3              18.0                    116.0    36.388889    83.5 NaN   
4              18.0                    116.0    36.388889    83.5 NaN   

  Patient_id  target  
0      30552       0  
1      30552       0  
2      30552       0  
3      30552       0  
4      30552       0

In [18]:
df.dtypes

Capillary refill rate                 float64
Diastolic blood pressure              float64
Fraction inspired oxygen              float64
Glascow coma scale eye opening         object
Glascow coma scale motor response      object
Glascow coma scale total              float64
Glascow coma scale verbal response     object
Glucose                               float64
Heart Rate                            float64
Height                                float64
Mean blood pressure                   float64
Oxygen saturation                     float64
Respiratory rate                      float64
Systolic blood pressure               float64
Temperature                           float64
Weight                                float64
pH                                    float64
Patient_id                             object
target                                  int64
dtype: object

Insights: 
- We have 14 fields of type float64
- 1 fields of type int64
- 4 fields of categorical data

In [19]:
df_dimensions = df.shape 
print(f"Dataset contains {df_dimensions[0]} rows and {df_dimensions[1]} columns." )

Dataset contains 300912 rows and 19 columns.


#### Descriptive statistics for Categorical data

In [22]:
df.describe(include="object")

Glascow coma scale eye opening Glascow coma scale motor response  \
count                          274190                            296978   
unique                              7                                12   
top                   4 Spontaneously                  6 Obeys Commands   
freq                            94516                            115595   

       Glascow coma scale verbal response Patient_id  
count                              296884     300912  
unique                                 12       6269  
top                          1.0 ET/Trach      11623  
freq                                87646         48

##### Get Summary of categorical data feature classes

In [63]:
def getColumnDetails(df):
    columns = df.select_dtypes(include=['object']).columns 
    column_summaries = {}
    for column in columns:
        unique_values = df[column].dropna().unique()
        num_of_unique = len(unique_values)
        column_summary = f"Feature: {column}, has {num_of_unique} unique features: {unique_values}"
        column_summaries[column] = column_summary
    for column, summary in column_summaries.items():
        print(summary) 

##### Columns Class Summary

In [62]:
getColumnDetails(df)

Feature: Glascow coma scale eye opening, has 8 unique features: ['Spontaneously' '1 No Response' '4 Spontaneously' '2 To pain'
 '3 To speech' nan 'To Pain' 'To Speech']
Feature: Glascow coma scale motor response, has 13 unique features: ['Obeys Commands' '4 Flex-withdraws' '5 Localizes Pain' 'No response' nan
 '6 Obeys Commands' '1 No Response' 'Localizes Pain' 'Flex-withdraws'
 'Abnormal extension' '2 Abnorm extensn' 'Abnormal Flexion'
 '3 Abnorm flexion']
Feature: Glascow coma scale verbal response, has 13 unique features: ['Oriented' '1.0 ET/Trach' nan 'No Response-ETT' '5 Oriented' '4 Confused'
 'Incomprehensible sounds' '2 Incomp sounds' '3 Inapprop words' 'Confused'
 '1 No Response' 'No Response' 'Inappropriate Words']
Feature: Patient_id, has 6269 unique features: ['30552' '6893_' '63456' ... '66532' '29378' '11623']


Insights: 

- **Glascow coma scale eye opening:**  
    - Has 274190 records
    - Nominal in nature
    - Composed of 7 unique classes
    - Most repeated frequency is '4 Spontaneously' with a frequency of 94516
    - Composed of 7 categories 'Spontaneously', '1 No Response', '4 Spontaneously', '2 To pain', '3 To speech', nan, 'To Pain', 'To Speech' since NaN is not a category but indicates missing values

- **Glascow coma scale motor response:**
    - Has 296978 records
    - Nominal in nature
    - Composed of 12 unique classes
    - Most repeated frequency is '6 Obeys Commands' with a frequency of 115595
    - Composed of 13 categories however NaN is not regarded as a category but indicates missing values

- **Glascow coma scale verbal response:**
    - Has 296884 records
    - Composed of 12 unique classes
    - Nominal in nature
    - Most repeated frequency is '1.0 ET/Trach' with a frequency of 87646
    - composed of  

- **Patient_id:**
    - Has 300912 records
    - Nominal in nature
    - Composed of 6269 unique classes/ patient_id's, implying that there existed 6269 patients in study
    - There might be a possibility of patients being assessed more than once
    - Most frequent patient has patiend_id of '11623' with a frequency of 48

Note: we do not have Binary categorical data

#### Descriptive statistics for Numerical data

In [17]:
df.describe(include="object")

Capillary refill rate  Diastolic blood pressure  \
count            6336.000000             296944.000000   
mean                0.219223                 62.541099   
std                 0.413753                341.559624   
min                 0.000000                  0.000000   
25%                 0.000000                 51.000000   
50%                 0.000000                 59.000000   
75%                 0.000000                 69.000000   
max                 1.000000             100105.010000   

       Fraction inspired oxygen  Glascow coma scale total        Glucose  \
count              88464.000000             184416.000000  300698.000000   
mean                   0.599884                 10.818123     130.628329   
std                    0.253919                  4.334923      84.171126   
min                    0.000000                  3.000000     -19.999974   
25%                    0.400000                  8.000000     101.000000   
50%                    0.500000                 11.000000     126.000000   
75%                    0.700000                 15.000000     158.000000   
max                    7.100000                 15.000000    9999.000000   

          Heart Rate        Height  Mean blood pressure  Oxygen saturation  \
count  300912.000000  55824.000000         296984.00000      300912.000000   
mean       79.447793    168.543422             78.79108          95.343460   
std        32.145920     15.137414             29.52986        2529.203751   
min       -19.999623      0.000000            -34.00000         -19.999687   
25%        70.000000    160.000000             68.00000          95.000000   
50%        84.000000    170.000000             77.00000          98.000000   
75%        97.000000    178.000000             88.00000         100.000000   
max       941.000000    203.000000           9381.00000      981023.000000   

       Respiratory rate  Systolic blood pressure    Temperature  \
count     300864.000000            300912.000000  298848.000000   
mean          18.731265               119.694213      36.832834   
std            6.884248                23.396042       1.000075   
min            0.000000                 0.000000       0.000000   
25%           15.000000               103.000000      36.277802   
50%           18.000000               117.000000      36.833333   
75%           22.000000               134.000000      37.388889   
max         1211.000000               295.000000      73.760002   

              Weight             pH         target  
count  221040.000000  230614.000000  300912.000000  
mean       82.969018       5.573617       0.142128  
std        26.765857       5.963634       0.349182  
min         0.000000     -19.999706       0.000000  
25%        66.600000       7.310000       0.000000  
50%        79.099998       7.370000       0.000000  
75%        94.699997       7.420000       0.000000  
max       931.224376      99.000000       1.000000

Insights: 

- **Capillary refill rate:**   
    - Has a 6336 data points
    - Most Data averages at around 0.219223 
    - Lowest value is 0 and highest value is 1 indicating presence of binary values probably presenting normal/ abnormal states
    - Has as standard deviation of 0.4138 
    - Most values are at 0, with some data points at 1, suggesting a skew towards normal states.

- **Diastolic blood pressure:**
    - Has 296944 data points recorded
    - Most levels fall around 62.54mmHg
    - Has a high standard deviation of 341.56, indicating high spread hence presence of outliers
    - the data will fall between 51 and 69 anything above or below is regarded as outliers
    - The max value is 100105 and min value is 0, indicating possible of error (possibly brought out during data collection phase)

- **Fraction inspired oxygen:**
    - Has 88464 data points recorded
    - Most levels fall around 0.599884
    - Has a standard deviation of 0.254, low spread hence little to no outliers present
    - the data will fall between 0.4 and 0.7 anything above or below is regarded as outliers
    - The max value is 7.1 and min value is 0, indicating possible presence of error or anomaly

- **Glascow coma scale total:**
    - Has 184416 records
    - Most levels fall around: 10.818123
    - Has a standard deviation of 4.334923
    - Lowest value is 3 and highest value is 15 
    - Median at 11, indicating that most patients fall in the mid-range of this scale

- **Glucose:**
    - Has 300698 records
    - Most Data averages at around 130.63 mg/dL
    - Has a high Standard deviation of 84.17, implying high variation hence possiblity of outliers
    - Lowest value is -19.99 and highest value is 9999 this indicates severe errors in data 
    - The values range between 101mg/dL to 158 mg/dL and anything outside and the range can be regarded as an outlier 

- **Heart Rate:**
    - Has 300912 records
    - Most rates average at around a mean of 79 beats per minute
    - Has a standard deviation of of 32.15 (moderate variability)
    - The least value is -19.999623 and maximum value is 941
    - based on the quatiles, most values fall between 70 and 97 beats per minutes(bpm) with a median value of 84 bpm
    - Note that there might exist errors in the data collected especially during data entry 

- **Height:**
    - Has 55824 records
    - Most heights average at around a mean of 168.54
    - Has a standard deviation of 15.14 which is not a very big deviation from the central points 
    - Most of the heights fall between 160 to 178 with a median of 170
    - Tallest individual had a height of 203 and shortest individual had a height of 0 (probable error here)

- **Mean blood pressure:**
    - Has 296984 records
    - Most levels average at around a mean of 78.79 mg/dL.
    - Has a standard deviation of 29.529 which is exudes a deviation from the central points 
    - Most of the levels fall between 68 and 88 with
    - Data set has a minimum value of -34 and max value of 9381, this is a possible data error

- **Oxygen saturation:**
    - Has 300912 records
    - Data averages with a mean of 95.343460
    - Has a standard deviation of 2529.203752, which is very high. This shows that there may exist outliers in the data
    - The Oxygen saturation levels fall between 95 and 100 with a central point of 98
    - The least saturation level recorded is -19.999687 and highest value of 981023(this is a possible error)

- **Respiratory rate:**
    - Has 300864 records
    - Data averages with a mean of 18.731265
    - Has as standard deviation of 6.884248(moderate variability), which is significantly ok hence data quality might be ok
    - The lowest value is 0 and largest being 1211(this may be erronoeus since the value is quite large)
    - The rates range between 15 and 22 with a median of 18

- **Systolic blood pressure:**
    - Has  300912 records
    - Data averages with a mean of  119.694213
    - Has as standard deviation of 23.396042(moderate variablity)
    - The lowest value is 0 and largest being 295
    - The rates range between 103 and 134 with a median of 117

- **Temperature:**
    - Has 298848 records
    - Data averages around 36.832834°C
    - Has as standard deviation of 1.000, which is small hence data quality might be ok
    - The lowest level is 0°C and largest being 73.76°C
    - The rates range between 36.277802°C and 37.388889°C with a median of 36.833333°C

- **Weight:**
    - Has 221040 records
    - Data averages with a mean of 82.969018
    - Has a standard deviation of 26.765857, which expresses that the data quality might not be ok
    - The lowest weight is at 0 and largest at 931.224376
    - The weights range between 66.600000 and 94.699997 with a median value of 79.099998

- **pH:**
    - Has 230614 records
    - Data averages with a mean of 5.573617
    - Has a standard deviation of 5.963634, which is not as high but expresses that the data quality might not be ok but at an insignificant level
    - The lowest weight is at -19.999706 and largest at 99
    - The weights range between 7.310000 and 7.420000 with a median value of 7.370000, the data seems to be compromised in term sof data quality because of the min and max values there exists outliers 



*Generally:*
- *Based on the above the data quality is challenged, as some features contain outliers and other contain missing values.*
    

#### Explorative Data Analysis
Approches of Exploratory Data Analysis (EDA)
*  Univariate Analysis
*  Bivariate Analysis
*  Multivariate Analysis

### Outlier Detection

In [64]:

df.plot(kind='box', subplots=True, layout=(2,2), figsize=(10, 8), title='Box Plots of Numerical Variables')
plt.show()

ImportError: matplotlib is required for plotting when the default backend "matplotlib" is selected.

### Relevant Functions 